### PR TITLE
[ROCm][DSv3.2] Adopt new paged-MQA-logits API + cached logits buffer with defensive padding

### PR DESCRIPTION
**Pull request by maeehart**
## Summary

Harden the ROCm sparse-MLA indexer for DeepSeek V3.2 decode on gfx950 (MI355X) against an intermittent HIP Memory Access Fault triggered by MTP speculative decoding.  Three changes, shipped together:

1. **New-API adoption.** Adopt the newer fused `deepgemm_fp8_paged_mqa_logits` aiter API (including the preshuffle path for `block_size == 64`) instead of the older 3-stage `_stage1` + Python `sum(dim=0)` pipeline.  This is required anyway for MFMA-shaped decode on gfx950 with `block_size > 1`; DSv3.2 already needs it and has been carrying the change as a private patch in its distribution image.
2. **Output-buffer caching.** Cache the output tensor across decode layers via a module-level `_get_paged_logits_buffer(...)`; DSv3.2 has 61 layers per decode step and the cache saves one `torch.full(-inf)` per layer.
3. **Defense-in-depth logits padding.** Over-allocate the cached buffer by `_PAGED_LOGITS_ROW_PADDING = 256` float32 columns.  Consumers see the logical shape; the downstream `top_k_per_row_decode` op already takes `stride(0)` / `stride(1)` as explicit arguments, so the padding is stride-transparent.

A companion aiter PR [ROCm/aiter#2866](https://github.com/ROCm/aiter/pull/2866) pairs with this one to add `mask=offset < max_model_len` to unmasked `buffer_store` sites in the preshuffle kernel as kernel hygiene; the exact root-cause kernel for the MAF has not been pinned yet, so this PR stands on its own empirical evidence and does not depend on aiter#2866 for correctness.

## Motivation

On gfx950 (MI355X), enabling MTP speculative decoding for DeepSeek V3.2 against `vllm-project/vllm` main built with a stock aiter reliably reproduces an HIP Memory Access Fault during decode.  A 20× MTP `c=4` sweep of the `(random_input_len=1000, random_output_len=100)` cell faulted on every run before these changes and completes 20 / 20 with zero MAFs after.

### What we know about the fault

- **Repro**: vLLM `main` + stock aiter + DSv3.2 + MTP `num_speculative_tokens >= 1` on MI355X.  Faults on every decode-heavy run without the padding; zero faults in 20 / 20 runs with it.
- **Fingerprint**: the reported faulting VA is consistently **2 MiB-aligned** — the hugepage size of HIP's caching allocator on current ROCm.  A 2 MiB-aligned fault is characteristic of a write that crosses into an adjacent (or unmapped) hugepage, rather than a simple integer overflow of an index expression inside a single bounds-checked op.
- **Causation**: not fully pinned yet.  A reasonable hypothesis — not proven — is that the padding changes the VA where `_cached_paged_logits` lands and moves a subsequent overshoot (possibly from a downstream PyTorch-Inductor-generated fused kernel, e.g. a `triton_poi_fused_*` whose Triton backend lowers stores to unchecked `global_store_dword`) away from the boundary of an adjacent allocation.  Reproducing the fault with `AMD_SERIALIZE_KERNEL=3 + AMD_LOG_LEVEL=4` on the unpatched image to attribute it to a single kernel name is on the follow-up list.

### Relationship to aiter#2866

The companion aiter PR [ROCm/aiter#2866](https://github.com/ROCm/aiter/pull/2866) adds `mask=offset < max_model_len` to the unmasked `buffer_store` sites in `_gluon_deepgemm_fp8_paged_mqa_logits_preshuffle[_varctx]`.  It pairs well with this padding as belt-and-suspenders kernel hygiene, but since the exact root-cause kernel for the MAF has not been conclusively identified, this PR does not rely on aiter#2866 to remove the fault — the +256 padding is proven on its own.

### Why the new-API adoption

Upstream `main` currently calls the older 3-stage `deepgemm_fp8_paged_mqa_logits_stage1` API, which allocates `(heads, B*next_n, max_model_len)` and sums over heads in Python — fine for correctness but leaves the fused/preshuffle kernel (required for MFMA shapes on gfx950 with `block_size > 1`) on the table.  The DSv3.2 distribution image has been carrying a patched `rocm_aiter_mla_sparse.py` that already uses the fused API; this PR upstreams that adoption so the MI355X decode path works out-of-the-box against mainline vLLM.

## Changes

**1. `_get_paged_logits_buffer(rows, cols, device)`** — a new module-private helper that returns a `(rows, cols)` float32 view initialised to `-inf`.  Internally caches an over-allocated `(rows, cols + _PAGED_LOGITS_ROW_PADDING)` tensor across decode steps: repeated calls with matching `(rows, cols, device)` reuse the same storage.

`_PAGED_LOGITS_ROW_PADDING = 256` (one float32 row pitch widening).

**2. `rocm_fp8_paged_mqa_logits(..., block_size: int = 1)`** — adds a `block_size` kwarg.  When `block_size > 1` **and** the installed aiter build exposes `deepgemm_fp8_paged_mqa_logits`, the fused path is used:

```python
out_logits = _get_paged_logits_buffer(batch_size * next_n, max_model_len, q_fp8.device)
_deepgemm_fp8_paged_mqa_logits(
    q_fp8, kv_cache_fp8, weights, out_logits,
    context_lens, block_tables, max_model_len,
    ChunkK=256, Preshuffle=(block_size == 64),
    KVBlockSize=block_size, WavePerEU=2,
)
return out_logits
```

Otherwise falls back to the existing `_stage1` + `out_qk.sum(dim=0)` path unchanged.  `block_size=1` (the default) preserves exactly today's behaviour on every existing codepath.

**3. Caller threads `block_size`.**  `rocm_aiter_sparse_attn_indexer` now passes `block_size=kv_cache.shape[1]` to `rocm_fp8_paged_mqa_logits`.  This is the same quantity already computed elsewhere in the file.

## Stride transparency (why the widened `stride(0)` is safe)

The only consumer of the logits returned by `rocm_fp8_paged_mqa_logits` in this file is `torch.ops._C.top_k_per_row_decode`, invoked immediately downstream:

```python
torch.ops._C.top_k_per_row_decode(
    logits, next_n, decode_metadata.seq_lens,
    topk_indices, num_rows,
    logits.stride(0),   # explicit
    logits.stride(1),   # explicit
    topk_tokens,
)
```

The C++ op (`large_context_topk` in `csrc/attention/topk.cu`) pulls `input_stride = score.stride(0)` into its `FastTopKParams` and only asserts `score.stride(1) == 1`.  Both are satisfied by our view (`stride(0) = cols + 256`, `stride(1) = 1`), so the widened-stride buffer is a fully legal input with no kernel change required.

## Cost

Per decode step (not per layer — the buffer is cached):

- **Memory**: `_PAGED_LOGITS_ROW_PADDING * batch * next_n * 4 B = 1 KiB × (batch × next_n)`.  For a typical DSv3.2 MTP config with `batch = 128, next_n = 2`, that's **~256 KiB extra VRAM total** (once per process).
- **Bandwidth**: unchanged — the kernel still writes only the logical `cols` per row.  The extra 256 columns per row are never read or written.
- **Wall clock**: none measurable — a one-time `torch.full` on shape change, amortised to zero across 61 DSv3.2 decode layers by the cache.

## Validation

- **Decode correctness on MI355X with MTP enabled** — 20 / 20 sweeps of a DSv3.2 MTP `num_speculative_tokens=1`, `max_concurrency=4` benchmark completed with zero MAFs against the same aiter build that was faulting on every run before the padding was added.  Speculative decoding functional (positive MTP acceptance metrics).  Serving stance: `--async-scheduling`, no `--enforce-eager`, `gpu-memory-utilization 0.9`, `--tensor-parallel-size 4`, `--block-size 64`, `--max-num-batched-tokens 16384` — i.e. the same production stance the DSv3.2 serving config ships with.
- **Legacy `_stage1` path unchanged** — `block_size=1` (the default) skips the new branch entirely, so every existing caller is byte-identical at the bytecode level to today's behaviour.

## Back-compat

- `block_size: int = 1` is a default-valued keyword arg — every existing caller is source- and ABI-compatible.
- When `block_size == 1` (or the installed aiter build doesn't export `deepgemm_fp8_paged_mqa_logits`), the code takes the same `_stage1` branch as before, with zero line-level behavioural change.
- The new helper is module-private (`_` prefix) and carries explicit documentation of its lifetime and the fault it guards against.

## Test plan

- [x] DSv3.2 MTP `c=4` decode on MI355X with a stock aiter build — 20 / 20 no-MAF sweep.
- [ ] DSv3.2 MTP `c=4` decode on MI355X with an aiter build that has #2866 applied — confirm numerical parity with the current path.
- [ ] Existing CI on the vLLM `_stage1` fallback path (`block_size = 1` default) — unchanged behaviour.
- [ ] Non-ROCm backends: untouched; the entire new branch is guarded by the aiter-module existence check.

## Cross-references

- aiter: https://github.com/ROCm/aiter/pull/2866  (in-kernel `buffer_store` mask on preshuffle)
- Internal tracking: DSv3.2 MTP Bug B (intermittent MAF on MI355X decode)

## Essential Elements of an Effective PR Description Checklist

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)."
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

**Comment by maeehart** [comment]
## Blocker: follow-up MAF reproduction on `num_speculative_tokens=2`

Moving this PR back to **draft** pending investigation of a follow-on finding.

### What the last push claims

The v2 body (just pushed) states the `+256` column padding eliminates the intermittent 2 MiB-aligned HIP MAF on MI355X / DSv3.2 MTP decode, backed by a 20 / 20 clean sweep.

### What that 20 / 20 actually covered

- `num_speculative_tokens = 1` (i.e. MTP drafts 1 token per step)
- `max_concurrency = 4`, `random_input_len = 1000`, `random_output_len = 100`
- Full production stance: `--async-scheduling`, no `--enforce-eager`, `gpu-memory-utilization 0.9`, `tensor-parallel-size 4`, `block-size 64`.

### What broke today

When I re-ran the same code path against the same image, same cell, same stance, but with `num_speculative_tokens = 2` (as part of an internal MTP benchmark sweep), decode faulted again:

```
Memory access fault by GPU node-6 (Agent handle: 0x1c8de190)
on address 0x7dd98ee00000. Reason: Unknown.
```

`0x7dd98ee00000` is exactly 2 MiB-aligned — the same fingerprint described in the PR motivation. The `+256` padding was confirmed loaded (`[BugB/G6] paged-logits buffer over-allocation INSTALLED (+256 cols padding)` printed by every worker), MTP was functional in the preceding requests (acceptance length 2.46, draft acceptance 73 %), and the fault hit 5 / 32 prompts into the first benchmark cell.

### What this changes about the PR's claim

The empirical story for `num_speculative_tokens = 1` is unchanged: 20 / 20 clean. The **stronger claim** in the PR body — that the padding is the fix for this MAF class in general — is *not* supported by the new evidence at `num_speculative_tokens = 2`. I don't want to land a claim I can't defend.

### What I'm going to do before re-opening

1. Reproduce the `n_spec = 2` fault N times to rule out a single-shot heisenbug.
2. Run the repro under `AMD_SERIALIZE_KERNEL=3 + AMD_LOG_LEVEL=4` on the unpatched image to pin the actual faulting kernel name, so we stop speculating about which kernel is the culprit.
3. Rebuild the test image against an aiter that carries [ROCm/aiter#2866](https://github.com/ROCm/aiter/pull/2866) (in-kernel `buffer_store` masks on the preshuffle path) and re-test `n_spec = 2` with both the padding and the kernel mask in place. If that clears, it's meaningful evidence that #2866 is (at least part of) the correct root-cause fix and not just kernel hygiene.
4. Rewrite the PR body with whichever of those three the evidence supports: narrow the scope to `n_spec = 1`, or promote `aiter#2866` back to a required prerequisite, or (if the Inductor/Triton `global_store` hypothesis is what survives) pivot the narrative entirely.

The code change itself (new-API adoption, output-buffer caching, +256 padding) is still almost certainly something we want — but the *why* in the description needs to match what the evidence actually supports. Back to draft until I have that evidence.

**Comment by maeehart** [comment]
## Current-stack re-evaluation

I completed the `num_speculative_tokens=2` follow-up from my blocker comment above. I am closing this PR as superseded.

I tested `deepseek-ai/DeepSeek-V3.2` on 4 x MI355X with TP=4, block size 64, vLLM `cb8104839c141609d99f1254459ef3a4f1bd4263`, and AITER `31350226161346314b3d8882c8085bd31dce6a34` (`0.1.19`). The public image was `vllm/vllm-openai-rocm@sha256:d22922d540810d90c5a3eafe91d3b4a62c2b881f3e990d0b7180b2875a5d176d`.

| Variant | Fresh starts | Trials | Requests | MAFs |
|---|---:|---:|---:|---:|
| Current masks, no padding | 4 | 23/23 | 736/736 | 0 |
| Current masks, +256 columns | 4 | 23/23 | 736/736 | 0 |
| Ten #2866 masks removed, no padding, serialized | 1 | 5/5 | 160/160 | 0 |
| Ten #2866 masks removed, no padding, `max_model_len=1100` boundary | 1 | 5/5 | 160/160 | 0 |

Each trial used 32 prompts, 1000 input tokens, 100 output tokens, concurrency 4, and MTP with `num_speculative_tokens=2`. Across 56 valid trials and 1,792 requests, no MAF occurred. The serialized control contained 4,544 dispatch records for `_gluon_deepgemm_fp8_paged_mqa_logits_preshuffle` and zero memory-access-fault records.

The +256 padding has no observed reliability benefit on the current stack. The #2866 masks are present through ROCm/aiter#3005, but removing them did not restore the historical failure in these bounded controls. Therefore, this result does not establish that the masks alone fixed the old MAF. No current faulting kernel can be named because no fault reproduced.

The useful non-padding work has landed independently through #41217, #42062, #41002, and #44527. The tested nightly also includes #49714, which fixes a distinct invalid-logits to invalid-sparse-index OOB path. Any future current-stack MAF should be investigated as a new issue with a fresh serialized kernel capture.
